BTIC_GBM.revised Normalization
========================================================
## Author: Sumaiya Islam
## Date: December 15, 2015

### A. Set up wd & packages

Here we will continue from pre-processing of our samples to the normalization of our samples using the objects BTIC_GBM.revised.RData and BTIC_GBM.revised.2.RData.

Initially we will set our working directory and load our libraries:
```{r BTIC_GBM.revised_setup, include = FALSE, echo=FALSE}
setwd("/home/sislam/BTIC_GBM")
library(methylumi)
library(gplots)
library(marray)
library(lumi)
library(lattice)
library("RColorBrewer")
library(knitr)
library(xtable)
library(wateRmelon)
library(limma)
library(sva)
library(RPMM)
```

### B. Load files

Only two files need to be loaded before we can begin:
- **BTIC_GBM.revised**, this is the MethylLumiM object which has been pre-processed and should contain 420,195 probes (in our particular case) and  samples.
- **BTIC_GBM.revised.2**, is the MethyLumiSet object which is needed for the normalization process and has not yet been utilized.

```{r BTIC_GBM.revised.rev_fileLoad, echo=FALSE}
load("BTIC_GBM.revised(filtered).RData")
load('BTIC_GBM.Rev.2.RData')
dim(BTIC_GBM.revised) # probes = 420,195, n = 109
dim(BTIC_GBM.revised.2) # probes = 485,577, n = 117
```
**BTIC_GBM.revised.rev** should contain 420,195 probes and 109 samples where **BTIC_GBM.revised.2** should contain 485,577 probes and 109 samples.

### C. Density Plots

Before we get into colour correction and normalization, we'll generate a density plot of the raw data from BTIC_GBM.revised.rev with the probes filtered. We will accordingly compare these raw density plots to the normalized density plots at the end of the script.

```{r BTIC_GBM.revised.rev_filteredDensity, echo=FALSE}
BatchCol <- rainbow(nlevels(factor(BTIC_GBM.revised$Batch)))[as.numeric(factor(BTIC_GBM.revised$Batch))]

density(BTIC_GBM.revised, col=BatchCol, lty = 1, legend=NULL, main = "BTIC_GBM.revised after probe filtering - M-values by Chip")
legend("topright",title="Batch", fill=rainbow(nlevels(factor(BTIC_GBM.revised$Batch))),legend=levels(factor(BTIC_GBM.revised$Batch)),cex=0.6)
```
You can see how much inter-sample variability exists across the batches. 

```{r BTIC_GBM.revised.rev_tissueDensity, echo=FALSE}
tissueCol <- as.numeric(factor(BTIC_GBM.revised$Sample_Group))
tissueCol <- gsub("1", "red",  gsub("2", "forestgreen", gsub("3", "blue", tissueCol)))
density(BTIC_GBM.revised, col=tissueCol, lty = 1, legend=NULL, main = "BTIC_GBM.revised after probe filtering- M-values by tissue")
legend("topleft", legend = c("Blood", "BTIC", "Tumour"),fill = c("red", "forestgreen", "blue"), horiz = T, cex=0.5)
```
You can see that the different tissue types have largely different methylation density profiles. Notably, the tumour and BTIC samples have considerable inter-sample variability, particularly at intermediate DNA methylation levels, over their blood counterparts. This may be due to aberrant DNA methylation in the tumour cells themselves as well as notable/gross differences in the cell types represented in the tumour (ie stromal cells, blood contamination, etc). 

```{r BTIC_GBM.revised.rev_SexDensity, echo=FALSE}
SexCol <- as.numeric(factor(BTIC_GBM.revised$Sex))
SexCol <- gsub("1", "deeppink", gsub("2", "dodgerblue", SexCol))
density(BTIC_GBM.revised, col=SexCol, lty = 1, legend=NULL, main = "BTIC_GBM.revised after probe filtering- M-values by Sex")
legend("topleft", legend = c("Female", "Male"),fill = c("deeppink","dodgerblue"), horiz = T, cex=0.7)
```
We observe limited variability in the methylation profiles that are driven by sex differences.

### D. Subset dataset

##### Important: We will subset the dataset by tissue source (ie blood, tumour and cell line and matched tumour/cell) to normalize separately. We will compare these individual normalized datasets to a dataset where all three subsets have been normalized together. 

```{r echo=FALSE}

# clone original filtered methylumi objects
all.dat<-BTIC_GBM.revised
all.dat.2<-BTIC_GBM.revised.2

# subset all cancer samples (only matched tumour and BTIC pairs)
cancer<-c("tumour", "cell")
cancer.samples <- grep(paste(cancer,collapse="|"), all.dat$Sample_Group)
cancer.dat<-all.dat[,cancer.samples] 
unmatched.samples<-c("76_cell", "189_cell", "42_cell", "48_cell", "166_cell", "12_cell", "12_cell_rep")
cancer.dat <- cancer.dat[,!(sampleNames(cancer.dat))%in%unmatched.samples]
dim(cancer.dat) # probes = 420,195 n = 73
cancer.dat.2<-all.dat.2[,cancer.samples] 
cancer.dat.2 <- cancer.dat.2[,!(sampleNames(cancer.dat.2))%in%unmatched.samples] 
dim(cancer.dat.2) # probes = 485,577 n = 73
identical(sampleNames(cancer.dat), sampleNames(cancer.dat.2))

# subset all blood samples
blood.samples <- grep("blood", all.dat$Sample_Group)
blood.dat<-all.dat[,blood.samples] 
dim(blood.dat) # probes = 420,195 n = 29
blood.dat.2<-all.dat.2[,blood.samples] 
dim(blood.dat.2) # probes = 485,577 n = 29
identical(sampleNames(blood.dat), sampleNames(blood.dat.2))

# subset all BTIC samples
BTIC.samples <- grep("cell", all.dat$Sample_Group)
BTIC.dat<-all.dat[,BTIC.samples] 
dim(BTIC.dat) # probes = 420,195 n = 45
BTIC.dat.2<-all.dat.2[,BTIC.samples] 
dim(BTIC.dat.2) # probes = 485,577 n = 45
identical(sampleNames(BTIC.dat), sampleNames(BTIC.dat.2))

# subset all tumour samples
tumour.samples <- grep("tumour", all.dat$Sample_Group)
tumour.dat<-all.dat[,tumour.samples] 
dim(tumour.dat) # probes = 420,195 n = 35
tumour.dat.2<-all.dat.2[,tumour.samples] 
dim(tumour.dat.2) # probes = 485,577 n = 35
identical(sampleNames(tumour.dat), sampleNames(tumour.dat.2))

# save(all.dat, file = "All.dat.raw.RData")
# save(cancer.dat, file = "Cancer.dat.raw.RData")
# save(blood.dat, file = "Blood.dat.raw.RData")
# save(BTIC.dat, file = "BTIC.dat.raw.RData")
# save(tumour.dat, file = "Tumour.dat.raw.RData")
```

### E. Quantro

The quantro R-package can be used to test for global differences between groups of distributions which assess whether global normalization methods such as quantile normalization should be applied.

```{r}
library(quantro)
matboxplot(as.matrix(betas(all.dat)), groupFactor = all.dat$Sample_Group, col = c(2,3,4), xaxt = "n",main = "Distribution of Methylation Values in All.dat", ylab="Beta Value", xlab="Samples") 
matboxplot(as.matrix(betas(cancer.dat)), groupFactor = cancer.dat$Sample_Group, col = c(3,4), xaxt = "n",main = "Distribution of Methylation Values in Cancer.dat", ylab="Beta Value", xlab="Samples")
matboxplot(as.matrix(betas(blood.dat)), groupFactor = blood.dat$Sample_Group, col = 2, xaxt = "n",main = "Distribution of Methylation Values in Blood.dat", ylab="Beta Value", xlab="Samples")
matboxplot(as.matrix(betas(BTIC.dat)), groupFactor = BTIC.dat$Batch, col = c(2,3), xaxt = "n",main = "Distribution of Methylation Values in BTIC.dat", ylab="Beta Value", xlab="Samples")
matboxplot(as.matrix(betas(tumour.dat)), groupFactor = tumour.dat$Sample_Group, col = 4, xaxt = "n",main = "Distribution of Methylation Values in Tumour.dat", ylab="Beta Value", xlab="Samples")
```

We can see differences in the median methylation signals between blood (red), tumours (blue) and BTIC samples (green). There is considerable inter-sample variability in the all.dat, cancer.dat, tumour.dat and BTIC.dat but not so much in the blood.dat

```{r}
library(doParallel)
registerDoParallel(cores=4)

# all.dat
(qtestPerm.all <- quantro(as.matrix(betas(all.dat)), all.dat$Sample_Group, B = 1000)) # quantroStat = 17.93457, quantroPvalPerm = 0 
quantroPlot(qtestPerm.all)

# cancer.dat 
cancer.groups<-as.character(cancer.dat$Sample_Group)
cancer.groups<-as.factor(cancer.groups)
(qtestPerm.cancer <- quantro(as.matrix(betas(cancer.dat)), cancer.groups, B = 1000)) # quantroStat = 8.67204, quantroPvalPerm = 0.003
quantroPlot(qtestPerm.cancer)

# blood.dat
(qtestPerm.blood <- quantro(as.matrix(betas(blood.dat)), blood.dat$Batch, B = 1000))  # quantroStat = 0.57575, quantroPvalPerm= 0.567
quantroPlot(qtestPerm.blood)

# BTIC.dat
(qtestPerm.BTIC <- quantro(as.matrix(betas(BTIC.dat)), BTIC.dat$Batch, B = 1000)) # quantroStat = 4.74169, quantroPvalPerm= 0.034
quantroPlot(qtestPerm.BTIC)

# tumour.dat
(qtestPerm.tumour <- quantro(as.matrix(betas(tumour.dat)), tumour.dat$Batch, B = 1000)) # quantroStat = 1.66999, quantroPvalPerm= 0.208
quantroPlot(qtestPerm.tumour)


#p<0.01 then there are global differences in distributions across groups
```

We can see that the quantro stat is significant for the all.dat and the cancer.dat (permuted p-value <0.01), indicating that there are global differences in distributions across groups so quantile normalization is not appropriate for these datasets. However, the blood.dat, BTIC.dat and tumour.dat does not have significant quantro stats so we can conclude that quantile normalization is appropriate for these datasets. 

### F. Quantile Normalization

```{r}
library(lumi)
blood.dat.q <- lumiMethyN(blood.dat, method = "quantile")
BTIC.dat.q <- lumiMethyN(BTIC.dat, method = "quantile")
tumour.dat.q <- lumiMethyN(tumour.dat, method = "quantile")
```


### G. Probe-type Normalization: BMIQ

The 450K Illumina Infinium Array has inherent variation associated with its methodologies which must be accounted for in our analyses. Much of this variation is attributed to the use of two types of probes used in the array, Type I and Type II.

Type I probes contain two bead types corresponding to an unmethylated (U) or methylated (M) status. Type I probes obtain methylation status as a result of fluoresence expressed after a single base pair extension occurs just after the target basepair, resulting in only one color channel being utilized (red). Type I probes also assume that any CpG sites underlying the probe are of the same status as the query site (methylated or unmethylated). The beta values for Type I probes are then determined by this formula b= M/(U + M). Importantly, Type I probes are enriched in regions of high CpG density (carry 3 or more CpG sites underlying the probe body), particularly those associated with promoters). Type II probes tend to occur in lower CpG density regions (carry 3 or less CpG sites underlying the probe body). Type II probes do not assume the methylation status of underlying CpG sites within the probe and so consist of a combination of degenerate probes with varying combinations of up to three underlying CpG sites. Type II probes also detect methylation status with a single base pair extension except that the site being extended is the CpG site of detection and so require two fluorescent colors green for methylated (M) and red for unmethylated (U) sites. Type II probe beta values are calculated using this formula b = Green (M)/(Red (U) + Green (M)). In terms of considerations for normalization, Type I probes have a much higher dynamic range than Type II probes, which may introduce an enrichment bias towards Type I probes when ranking probes in supervised analyses (Teschendorff et al 2013, Bioinformatics). 

Due to these inherent differences between Type I and Type II probes used in the Illumina Infinium 450K array several groups in the field have deveolped various normalization analyses to correct for the differences between these probes. We will be using an intra-sample normalization method that corrects for probe-type differences called BMIQ (Beta MIxture Quantile dilation) (Teschendorff et al 2013, Bioinformatics).

#### BMIQ (Beta Mixture Quantile dilation)

BMIQ is an intra-sample normalisation procedure, correcting the bias of type-2 probe values. BMIQ uses a 3-step procedure: (i) fitting of a 3-state beta mixture model, (ii) transformation of state-membership probabilities of type2 probes into quantiles of the type1 distribution, and (iii) a conformal transformation for the hemi-methylated probes. Exact details can be found in the reference (Teschendorff et al 2013, Bioinformatics).


```{r BTIC_GBM.revised.rev_BMIQ, echo=FALSE}
# define probe design vector
head(probe_design<-as.character(fData(all.dat)$INFINIUM_DESIGN_TYPE))
probe_design.v<- replace(probe_design, probe_design=="I", 1)
probe_design.cor<- replace(probe_design.v, probe_design.v=="II", 2)
probe_design.cor<-as.numeric(probe_design.cor)
identical(nrow(exprs(all.dat)), length(probe_design.cor))

# Run BMIQ across each dataset (this code takes a long time---best to run it overnight esp if you have a lot of samples)
all.dat_normalized<-apply(betas(all.dat), 2, function(x) BMIQ(x,probe_design.cor)) # all.dat
cancer.dat_normalized<-apply(betas(cancer.dat), 2, function(x) BMIQ(x,probe_design.cor)) # cancer.dat
blood.dat_normalized<-apply(betas(blood.dat.q), 2, function(x) BMIQ(x,probe_design.cor)) # blood.dat
BTIC.dat_normalized<-apply(betas(BTIC.dat.q), 2, function(x) BMIQ(x,probe_design.cor)) # BTIC.dat
tumour.dat_normalized<-apply(betas(tumour.dat.q), 2, function(x) BMIQ(x,probe_design.cor)) # tumour.dat


# extract normalized beta values and reshape
reshape.BMIQ.betas<-function(bmiq.output,methy.object){
  bb<-lapply(1:length(bmiq.output), function(x) bmiq.output[[x]]$nbeta)
  cc<-do.call(rbind, bb)
  dd<-t(cc)
  colnames(dd)<-colnames(exprs(methy.object))
  ee<-methy.object
  mvals<-Beta2M(dd)
  exprs(ee)<-mvals
  return(ee)
}

## reshape BMIQ output into new methylumi object for each dataset
all.dat.BMIQ<-reshape.BMIQ.betas(all.dat_normalized,all.dat)
cancer.dat.BMIQ<-reshape.BMIQ.betas(cancer.dat_normalized, cancer.dat)
blood.dat.BMIQ<-reshape.BMIQ.betas(blood.dat_normalized,blood.dat.q)
BTIC.dat.BMIQ<-reshape.BMIQ.betas(BTIC.dat_normalized,BTIC.dat.q)
tumour.dat.BMIQ<-reshape.BMIQ.betas(tumour.dat_normalized,tumour.dat.q)


# save normalized datasets 
# save(all.dat.BMIQ, file="All.Dat.BMIQ.RData")
# save(cancer.dat.BMIQ, file="Cancer.Dat.BMIQ.RData")
# save(blood.dat.BMIQ, file="Blood.Dat.BMIQ.RData")
# save(BTIC.dat.BMIQ, file="BTIC.Dat.BMIQ.RData")
# save(tumour.dat.BMIQ, file="Tumour.Dat.BMIQ.RData")
```

## Comparing raw and normalized datasets

For all samples (PBMC samples are in yellow while buccal samples are in blue):
```{r BTIC_GBM.revised.rev_compareAll, echo=FALSE}
# par(mfrow = c(1, 2))
# density(BTIC_GBM.revised.rev,legend=NULL, lty = 1, xlim = c(-6, 6), ylim=c(0,0.2), main = "Raw", col=tissueCol)
# density(BTIC_GBM.revised.rev.BMIQ,legend=NULL, lty = 1, xlim = c(-6, 6), ylim=c(0,0.2), main = "BMIQ", col=tissueCol)

## use ggplot2 to plot density plots of raw and normalized data
# extract beta matrix for raw and normalized datasets
density.plot.raw.norm<-function(x, y){
  betas.raw<-betas(x)
  betas.norm<-betas(y)
  # randomnly sample 10,000 probes from the datasets (same random probes in each dataset)
  random.probes<-sample(1:nrow(betas.raw), 10000)
  betas.raw.subset<-betas.raw[random.probes,]
  betas.norm.subset<-betas.norm[random.probes,]
  # melt each dataset
  head(betas.raw.melt<-melt(betas.raw.subset))
  head(betas.norm.melt<-melt(betas.norm.subset))
  # remove NAs
  betas.raw.melt.clean<-betas.raw.melt[which(!(is.na(betas.raw.melt$value))),]
  betas.norm.melt.clean<-betas.norm.melt[which(!(is.na(betas.norm.melt$value))),]
  # add descriptor for each datatype and add meta data to each dataset
  betas.raw.melt.clean$Data<-"Raw"
  betas.norm.melt.clean$Data<-"Normalized"
  head(betas.raw.melt.clean)
  meta.raw<-pData(x)
  meta.norm<-pData(y)
  meta.raw$SampleName<-rownames(meta.raw)
  meta.norm$SampleName<-rownames(meta.norm)
  head(betas.raw.plot<-merge(betas.raw.melt.clean, meta.raw, by.x="Var2", by.y="SampleName"))
  head(betas.norm.plot<-merge(betas.norm.melt.clean, meta.norm, by.x="Var2", by.y="SampleName"))
  # combine both datasets
  betas.plot<-rbind(betas.raw.plot, betas.norm.plot)
  #re-order levels of Data variable
  betas.plot$Data<-factor(betas.plot$Data, levels=c("Raw", "Normalized"))
  # plot density plots
  ggplot(betas.plot, aes(value, group=Var2, colour=Sample_Group)) + geom_density() + theme_bw() + facet_wrap(~Data) 
}

density.plot.raw.norm(all.dat, all.dat.BMIQ)
density.plot.raw.norm(cancer.dat, cancer.dat.BMIQ)
density.plot.raw.norm(blood.dat, blood.dat.BMIQ)
density.plot.raw.norm(BTIC.dat, BTIC.dat.BMIQ)
density.plot.raw.norm(tumour.dat, tumour.dat.BMIQ)
```

We can see that there is reduced sample-to-sample variability in the BMIQ-normalized dataset versus the raw data. 

### For the purposes of time efficiency, we will carry out all subsequent steps using just the cancer.dat.BMIQ and BTIC.dat.BMIQ datasets as these datasets are most relevant to our intended analyses. 

## Check the technical replicate correlation before and after normalization
```{r}

RepCor  <- function(x,y) {
  sRep <- x[, sampleNames(x)%in%y]
  cor.dat<- cor(betas(sRep), use = "pairwise.complete.obs")
  return(result = cor.dat[1, 2])
}

##  define within batch technical replicates
Rep067<-c("67_cell", "67_cell_rep")
Rep147<-c("147_cell", "147_cell_rep")

##  define across batch technical replicates
Rep073<-c("73_cell", "73_cell_rep")
Rep012<-c("12_cell", "12_cell_rep") # only present in BTIC.dat

tech.rep.comparisons<-function(raw.dat, norm.dat){
  ## compute correlations for raw cancer.dat
  # Within Batch 1  technical replicates
  cor067.raw<-RepCor(raw.dat, Rep067) 
  cor147.raw<-RepCor(raw.dat, Rep147) 
  # avg.batch1.raw<-mean(c(cor067.raw, cor147.raw))
  # across batch technical replicates
  cor073.raw<-RepCor(raw.dat, Rep073)
  cor012.raw<-RepCor(raw.dat, Rep012) 
  # avg.cross.batch.raw<-mean(c(cor073.raw, cor012.raw))
  ## compute correlations for raw cancer.dat
  # Within Batch 1  technical replicates
  cor067.norm<-RepCor(norm.dat, Rep067) 
  cor147.norm<-RepCor(norm.dat, Rep147) 
  # avg.batch1.norm<-mean(c(cor067.norm, cor147.norm))
  # across batch technical replicates
  cor073.norm<-RepCor(norm.dat, Rep073)
  cor012.norm<-RepCor(norm.dat, Rep012) 
  # avg.cross.batch.norm<-mean(c(cor073.norm, cor012.norm))
  ## generate output table
  Tech.Rep.Correlations<-data.frame(Before_BMIQ=c(cor067.raw, cor147.raw, cor073.raw, cor012.raw), After_BMIQ=c(cor067.norm, cor147.norm, cor073.norm, cor012.norm))
  rownames(Tech.Rep.Correlations)<-c("67 Rep", "147 Rep", "73 Rep", "12 Rep")
  return(Tech.Rep.Correlations)
}

(BTIC.dat.tech.rep.cor<-tech.rep.comparisons(BTIC.dat, BTIC.dat.BMIQ))


# for cancer.dat, need to omit all lines pertaining to Rep012 (this set, 12_cell and 12_cell_rep does not exist in the cancer.dat dataset
tech.rep.comparisons<-function(raw.dat, norm.dat){
  ## compute correlations for raw cancer.dat
  # Within Batch 1  technical replicates
  cor067.raw<-RepCor(raw.dat, Rep067) 
  cor147.raw<-RepCor(raw.dat, Rep147) 
  # avg.batch1.raw<-mean(c(cor067.raw, cor147.raw))
  # across batch technical replicates
  cor073.raw<-RepCor(raw.dat, Rep073)
  # cor012.raw<-RepCor(raw.dat, Rep012) 
  # avg.cross.batch.raw<-mean(c(cor073.raw, cor012.raw))
  ## compute correlations for raw cancer.dat
  # Within Batch 1  technical replicates
  cor067.norm<-RepCor(norm.dat, Rep067) 
  cor147.norm<-RepCor(norm.dat, Rep147) 
  # avg.batch1.norm<-mean(c(cor067.norm, cor147.norm))
  # across batch technical replicates
  cor073.norm<-RepCor(norm.dat, Rep073)
  # cor012.norm<-RepCor(norm.dat, Rep012) 
  # avg.cross.batch.norm<-mean(c(cor073.norm, cor012.norm))
  ## generate output table
  Tech.Rep.Correlations<-data.frame(Before_BMIQ=c(cor067.raw, cor147.raw, cor073.raw), After_BMIQ=c(cor067.norm, cor147.norm, cor073.norm))
  rownames(Tech.Rep.Correlations)<-c("67 Rep", "147 Rep", "73 Rep")
  return(Tech.Rep.Correlations)
}

(cancer.dat.tech.rep.cor<-tech.rep.comparisons(cancer.dat, cancer.dat.BMIQ))

```
We can see that the within-batch technical replicate correlations (67 Rep and 147 Rep) improved in both the cancer.dat and BTIC.dat after normalization. In addition, the across batch technical replicate 12_Rep in the BTIC.dat showed improved correlation after normalization. It is likely that these replicates represent true technical replicates (ie same bsDNA of that sample applied on different chips across batches). However, the across batch replicate 73_Rep showed low correlation (0.922) prior to normalization, which worsened after BMIQ normalization (0.921). It is possible that 73_Rep represents biological replicates (ie different collections of cell from different passages; note that their SNP profiles match so they are from the same individual). If it is the case, then this is indicative of a high degree of cell-culture-induced heterogeneity in the DNA methylation profiles of the BTICs. 

We will run ComBat on these datasets to minimize batch-to-batch effects. 


### F. Plot beta values before/after normalization

We will show the distribution of beta values for the entire dataset (all samples normalized together) in its raw form and final normalized form.

```{r BTIC_GBM.revised.rev_betaPlot, echo=FALSE} 
type <- fData(cancer.dat)$INFINIUM_DESIGN_TYPE

# for cancer.dat
par(mfrow = c(1, 1))
plot(c(0, 1) , c(0, 10), type = "n", xlab = "Beta values", ylab = "Density")
lines(density(na.omit(betas(cancer.dat))[type=="I",]),col="blue",lty=1)
lines(density(na.omit(betas(cancer.dat))[type=="II",]),col="blue",lty=2, wld=1.5)
lines(density(na.omit(betas(cancer.dat.BMIQ))[type=="I",]),col="red",lty=1)
lines(density(na.omit(betas(cancer.dat.BMIQ))[type=="II",]),col="red",lty=2, lwd=1.5)
legend("top", inset=.05, cex=0.6, c("Raw Type 1","Raw Type 2", "Normalized Type 1", "Normalized Type 2"), col=c("blue","blue","red", "red"), lty=c(1,2,1,2), horiz=FALSE)

# for BTIC.dat
par(mfrow = c(1, 1))
plot(c(0, 1) , c(0, 10), type = "n", xlab = "Beta values", ylab = "Density")
lines(density(na.omit(betas(BTIC.dat))[type=="I",]),col="blue",lty=1)
lines(density(na.omit(betas(BTIC.dat))[type=="II",]),col="blue",lty=2, wld=1.5)
lines(density(na.omit(betas(BTIC.dat.BMIQ))[type=="I",]),col="red",lty=1)
lines(density(na.omit(betas(BTIC.dat.BMIQ))[type=="II",]),col="red",lty=2, lwd=1.5)
legend("top", inset=.05, cex=0.6, c("Raw Type 1","Raw Type 2", "Normalized Type 1", "Normalized Type 2"), col=c("blue","blue","red", "red"), lty=c(1,2,1,2), horiz=FALSE)
```

We can see that there is improved peak-to-peak overlap of the Type 1 and Type 2 probes with the final normalized datasets of cancer.dat and BTIC.dat as desired (ie Type 1 and 2 probes have more similar dynamic ranges for their beta values). 

We will continue on with our analysis using the BMIQ-normalized datasets.