BTIC_GBM_Xeno Preprocessing 
========================================================
## Author: Sumaiya Islam
## Date: October 3, 2017

### A. Set up wd & packages

We will initially set our working directory and load our libraries.
```{r include = FALSE, echo=FALSE}
library(methylumi)
library(gplots)
library(marray)
library(lumi)
library(lattice)
library("RColorBrewer")
library(knitr)
library(xtable)
library(wateRmelon)
library(limma)
library(sva)
library(RPMM)
```

The **methylumi** package is required for loading and manipulating the Illumina methylation data. It is built to contain sample information, feature information, and multiple matrices of data (beta and transformed M-values) along with background correction and quality control features.

The **lumi** package is also important for processing Illumina 27K and 450K methylation microarray data. 

### B. Load files

Four files are required for our analysis:
- **sampleFile** contains all the sample information and variables. It can be produced in Excel. The variables contained in our study include tissue, sex, age of onset, Sentrix ID (chip barcode) and Sentrix Position (row and column position on the chip), etc. The sampleFile must contain a "SampleID" column which is the same as sample ID, except without "rep" for replicates in order to utilize the methylumi and lumi packages.
- **allFile** is generated in GenomeStudio.  It contains all probe annotation information (both Illumina's and reannotation file produced in Price et al. 2013 *Epigenetics*), and the following subcolumns for all samples: average beta, detection Pval, signal A, and signal B
- **betaFile** is generated in GenomeStudio.  It contains all probe annotation information, and all samples with only the average beta subcolumn.
- **qcFile** is generated in GenomeStudio (under the Control Probes profile).  It shows all samples (i.e. all columns), plus subcolumns: signal Grn, signal Red, and detection Pval.

### Note: Quality control steps for colour correction and background subtraction have been performed in GenomeStudio using invariant control probes. The corrected data were exported in the above-mentioned files. 

```{r echo=FALSE}
sampleFile <- ("BTIC_GBM_Xeno_Samplesheet_Oct_2017.txt")
allFile <- ("BTIC_GBM_Xeno_alldata.txt")
betaFile <- ("BTIC_GBM_Xeno_avgbeta.txt")
qcFile <- ("BTIC_GBM_Xeno_qcfile.txt")

file.exists(allFile)
file.exists(betaFile)
file.exists(qcFile)
file.exists(sampleFile)
```

Next, we'll make the BTIC_GBM_Xeno project (and BTIC_GBM_Xeno.2):

```{r echo=FALSE}
BTIC_GBM_Xeno<- lumiMethyR(allFile)
BTIC_GBM_Xeno.2 <- methylumiR(betaFile, qcfile = qcFile)


BTIC_GBM_Xeno # probes = 485,577, n = 12
BTIC_GBM_Xeno.2 # probes = 485,577, n = 12
```

The function *lumiMethyR()* coerces the object (**allFile**) into a **MethyLumiM class object**, which contains those four elements (exprs, methylated, unmethylated, and detection p value).

The function *methylumiR()* coerces the object (**betaFile**) into a **MethyLumiSet class object**, which holds the intensities and the beta values.  This is also where the sample information (the sampleFile) can be added as phenoData.
BTIC_GBM_Xeno.2 contains the information from the control probes, as well as the average beta values for all probes.

### C. Add metadata to BTIC_GBM_Xeno

```{r echo=FALSE}
# load and check sample names of samplesheet
sampleFile <- ("BTIC_GBM_Xeno_Samplesheet_Oct_2017.txt") 

str(sampleInfo <- read.delim(sampleFile,row.names=1,stringsAsFactors=F))
stopifnot(all(sampleNames(BTIC_GBM_Xeno)%in%rownames(sampleInfo)))


# re-order revised samplesheet according to sample name order of original samplesheet
sampleInfo.rev.or<-sampleInfo.rev[match(rownames(sampleInfo), rownames(sampleInfo.rev)),]
identical(rownames(sampleInfo), rownames(sampleInfo.rev.or)) # TRUE

# re-structure revised samplesheet's variables
sampleInfo <- sampleInfo[sampleNames(BTIC_GBM_Xeno),]
str(sampleInfo)
sampleInfo$Sample_Group<-as.factor(sampleInfo$Sample_Group)
sampleInfo$Sentrix_ID<-as.factor(sampleInfo$Sentrix_ID)
sampleInfo$Batch<-as.factor(sampleInfo$Batch)
sampleInfo$Sex<-as.factor(sampleInfo$Sex)

# input revised samplesheet in pData of BTIC_GBM_Xeno methylumi object
str(pData(BTIC_GBM_Xeno) <- sampleInfo)
```

First, we've read the sampleFile and made it into a matrix.  The first row of sampleFile becomes the row names of sampleInfo.  The **stringsAsFactors** argument specifies whether R should keep character variables as they are (= FALSE), or to convert them to factors (= TRUE).

We then perform a sanity check to ensure all BTIC_GBM_Xeno sample names are equivalent to sampleInfo row names.
Next, the sampleInfo is reordered to match the sample names of BTIC_GBM_Xeno. This is important because we are merging sampleInfo and BTIC_GBM_Xeno: sampleInfo becomes BTIC_GBM_Xeno's pData (or phenoData, i.e. the phenotypic information about samples in BTIC_GBM_Xeno)

```{r echo=FALSE}
BTIC_GBM_Xeno <- BTIC_GBM_Xeno[, order(BTIC_GBM_Xeno$Sentrix_ID)]
```

We chose to re-order BTIC_GBM_Xeno by Sentrix ID (array chip barcode number), but this could be done for other variables.

We will now subset our methylumi object based on tissue type to generate BTIC_GBM_Xeno and BTIC_GBM_Xeno. These objects will be examined and utilized in subsequent analyses.

### D. Saving BTIC_GBM_Xeno and BTIC_GBM_Xeno.2

```{r echo=FALSE}
# save(BTIC_GBM_Xeno, file ='BTIC_GBM_Xeno.RData')
# save(BTIC_GBM_Xeno.2, file ='BTIC_GBM_Xeno.2.RData')
```

### E. Preliminary plotting and exploratory analysis

We will generate some preliminary exploratory plots of the data to check for possible outliers or batch effects.

First we plot the sample relations based on MDS (multidimensional scaling) in order to visually investigate the data. We initially color the data according to Sentrix ID in order to assess possible batch effects. We then do a similar assessment, coloring the samples by Sample_Group (which is the tissue source) to investigate any obvious differences from our main variable of interest:

```{r fig.width=8, fig.height=7, echo=FALSE}
par(cex = 0.7)
plotSampleRelation (BTIC_GBM_Xeno,method = "mds", col = as.character (as.numeric (factor(BTIC_GBM_Xeno$Sentrix_ID))))
```

We can see that there doesn't appear to be much of a chip-to-chip effect.  

```{r fig.width=8, fig.height=7, echo=FALSE}
par(cex = 0.7)
plotSampleRelation (BTIC_GBM_Xeno,method = "mds", col = as.character (as.numeric (factor(BTIC_GBM_Xeno$Batch))))
```

We also observe minimal batch effects.

```{r fig.width=8, fig.height=7, echo=FALSE}
tissueCol <- as.numeric(factor(BTIC_GBM_Xeno$Sample_Group))
tissueCol <- gsub("1", "red",  gsub("2", "forestgreen", gsub("3", "blue", tissueCol)))

par(cex = 0.7)
plotSampleRelation(BTIC_GBM_Xeno, method = "mds", col = tissueCol)
legend("top", legend = c("BTIC", "Tumour", "Xeno"),fill = c("red", "forestgreen", "blue"), horiz = T)
```

It is apparent that tissue differences (ie blood vs tumour vs BTIC cell line) comprise a substantial proportion of the variance in the pre-processed methylation data (specified by principal component 1). This is expected given that different cell types are epigenetically distinct and accordingly, exhibit differential methylation profiles. 

We also produce a dendogram of the samples as another way to visualize possible tissue/cell effects.

```{r fig.width=8, fig.height=7, echo=FALSE}
par(cex = 0.7)
plotSampleRelation(BTIC_GBM_Xeno, method = "cluster")
```



```{r fig.width=8, fig.height=7, echo=FALSE}
sentrixCol <- as.character (as.numeric(factor(BTIC_GBM_Xeno$Sentrix_ID)))


BuPu <- colorRampPalette(brewer.pal(n = 9, "BuPu"))

cor <- cor(betas(BTIC_GBM_Xeno), use = "pairwise.complete.obs")

heatmap.2(cor,main = "Correlation of samples - 485,577 probes",
          trace = "none", col = BuPu, dendrogram = "both",
          RowSideColors = sentrixCol, cexRow = 1,
          ColSideColors = tissueCol, cexCol = 1,
          keysize = 1)
legend("topright", legend = c("BTIC", "Tumour", "Xeno"),fill = c("red", "forestgreen", "blue"), horiz = T)
legend ("topleft", legend = levels (factor (BTIC_GBM_Xeno$Sentrix_ID)), fill = 1:nlevels (factor (BTIC_GBM_Xeno$Sentrix_ID)), horiz = T)
```



```{r fig.width=8, fig.height=7, echo=FALSE}
batchCol <- as.numeric(factor(BTIC_GBM_Xeno$Batch))
batchCol <- gsub("1", "orange",  gsub("2", "purple", batchCol))

BuPu <- colorRampPalette(brewer.pal(n = 9, "BuPu"))

cor <- cor(betas(BTIC_GBM_Xeno), use = "pairwise.complete.obs")

heatmap.2(cor,main = "Correlation of samples - 485,577 probes",
          trace = "none", col = BuPu, dendrogram = "both",
          RowSideColors = batchCol, cexRow = 1,
          ColSideColors = tissueCol, cexCol = 1,
          keysize = 1)
legend("topright", legend = c("Blood", "BTIC", "Tumour"),fill = c("red", "forestgreen", "blue"), horiz = T)
legend ("topleft", legend = levels (factor (BTIC_GBM_Xeno$Batch)), fill = 1:nlevels (factor (BTIC_GBM_Xeno$Batch)), horiz = T)
```
 

### F. Probe Filtering

##### Removal of SNP Probes

Here we create a sub-project, **BTIC_GBM_Xeno.rs**, containing only information for the probes that are on single-nucleotide polymorphisms (SNPs (n = 65)).  We then remove these probes from **BTIC_GBM_Xeno**. We remove the SNP probes as they are used as an internal control to ensure your samples are what you think they are and are not used for any methylation analysis.

```{r echo=FALSE}
BTIC_GBM_Xeno.rs <- BTIC_GBM_Xeno[substring(featureNames(BTIC_GBM_Xeno), 1, 2) == "rs", ] 
dim(BTIC_GBM_Xeno.rs) # SNP probes = 65, n = 12
BTIC_GBM_Xeno <- BTIC_GBM_Xeno[substring(featureNames(BTIC_GBM_Xeno),1,2) != "rs", ]
dim(BTIC_GBM_Xeno) # probes = 485,512, n = 12
```

This cluster plot observes the samples based on the 65 SNP probes alone (i.e. uses the BTIC_GBM_Xeno.rs project)

```{r fig.width=12, fig.height=10, echo=FALSE}
plotSampleRelation(BTIC_GBM_Xeno.rs,method="cluster", cex=0.7)
```


##### Removal of XY Probes

Now we do the same thing for probes located on the X and Y chromosomes.  The mini project **BTIC_GBM_Xeno.xy** will contain only 11,648 probes; these are the probes we will remove from **BTIC_GBM_Xeno**. We remove probes located on the X and Y chromosome in this analysis because unlike autosomes, sex chromosomes are not in equal number between females (XX) and males (XY) and if your cohort is not sex matched you will have a disproportionate number of X vs Y chromosomes present in your analysis throwing off the data of those probes.

```{r echo=FALSE}
BTIC_GBM_Xeno.xy <- BTIC_GBM_Xeno[fData(BTIC_GBM_Xeno)$CHR%in%c("X", "Y"), ]
dim(BTIC_GBM_Xeno.xy) # probes = 11,648, n = 12
BTIC_GBM_Xeno <- BTIC_GBM_Xeno[!fData(BTIC_GBM_Xeno)$CHR%in%c("X", "Y"), ]
dim(BTIC_GBM_Xeno) # probes = 473,864, n = 12
```


```{r fig.width=8, fig.height=7, echo=FALSE}
SexCol <- as.numeric(factor(BTIC_GBM_Xeno$Sex))
SexCol <- gsub("1", "dodgerblue",  gsub("2", "deeppink", SexCol))
par(cex = 0.7)
plotSampleRelation(BTIC_GBM_Xeno.xy, method = "mds", col = SexCol)
legend("top", legend = c("Female", "Male"),fill = c("deeppink","dodgerblue"), horiz = T)
```


### H. Filtering out "bad" probes

```{r echo=FALSE, include=FALSE}
all(featureNames(BTIC_GBM_Xeno)%in%featureNames(BTIC_GBM_Xeno.2))
all(sampleNames(BTIC_GBM_Xeno)%in%sampleNames(BTIC_GBM_Xeno.2))
```


##### Missing beta values

We are looking for probes where there was no average beta value (the measure of methylation) calculated by GenomeStudio. This is done when GenomeStudio detects intensity values from less than three beads for a given probe. There are multiple beads on the array as a means of obtaining technical replicates and so probes which have fewer than three beads representing it is called as an NA.
**avgbeta** is a matrix of the beta values for each sample at each probe. 
**badAvgbeta** is a logical matrix that shows which probes have NAs for which samples.
From this we make **nbadAvgbeta**, which sums up the number of "TRUE"'s for each row.  Our threshold is 1 badAvgbeta for a probe to be removed.

```{r fig.width=8, fig.height=7, echo=FALSE}
avgbeta <- betas(BTIC_GBM_Xeno.2)[featureNames(BTIC_GBM_Xeno),sampleNames(BTIC_GBM_Xeno)]
badAvgbeta <- is.na(avgbeta)
nbadAvgbeta <- print(sum(rowSums(badAvgbeta)>=1))  #2,587

dotchart(colSums(badAvgbeta), cex=0.8, main = paste("Bad Avg Beta"))
```

This dot chart is shows how many probes are defined as having a "bad average beta value" by the threshold we specified above. As a general rule of thumb a sample has to have more than 10% (40,000) of probes deemed "bad" before it is considered for removal. Our worst sample has only ~500 probes with a bad average beta value. Here we have identified 2,587 probes with a bad average beta value using thresholds defined above. 

##### Bad detection p-values

**Detection p-values:** the confidence that the signal from the probe in question is more than the background (as defined by the negative control probes on the Illumina array).  We set our p-value threshold to 0.01.

First, we make the object **badDetP**, which is a logical matrix that lets you know which probes do not meet our p-value requirements for which samples (i.e. it is a 473,684 by 109 matrix).
From this we make **nbadDetP**, which sums up the number of "TRUE"'s in each row (i.e. for each probe).  We have set a threshold of 1 badDetP (detection p-value >= 0.01) to label a probe as a "bad" probe - this means that any one probe had to not meet our p-value threshold for at least 1 of our 109 samples to be targeted for removal.

```{r fig.width=8, fig.height=7, echo=FALSE}
badDetP <- detection(BTIC_GBM_Xeno)>0.01
nbadDetP <- print(sum(rowSums(badDetP)>=1)) #1907

dotchart(colSums(badDetP),main=paste("Bad Det P"), cex = 0.4)
```

As stated above a sample has to contain about 10% of bad probes to consider removing it. Our worst sample has only ~600. Here we have identified 6754 probes which have a bad detection P-value using thresholds defined above.


##### Removing these bad probes

```{r echo=FALSE}
badProbes <- rowSums(badAvgbeta)>=1|rowSums(badDetP)>=1
sum(badProbes)# badProbes = 4450
BTIC_GBM_Xeno <- BTIC_GBM_Xeno[!badProbes,] 
dim(BTIC_GBM_Xeno) # probes = 469,414, n = 12
```

We then take out all probes identified above as having either a bad detection P-value or a bad average beta value or both. In total we have removed 4450 "bad" probes from our study. 2587 (badAvgbeta) + 1907 (badDetP) = 4494 yet we removed only 4450. This is because there were some probes identified which overlapped as having both bad average beta values and bad detection P-values. We have a total of 469,414 probes left in our analysis.

##### Cross-hybridizing probes

Some probes have been found to cross-hybridize with other chromosomes (Price et al. 2013 *Epigenetics*).  It is at the discretion of the user whether or not they want to remove these cross-hybridizing probes, since it isn't a guarantee that they will cross-hybridize every time the experiment is run.  Probes that cross-hybridize to the sex chromosomes are typically removed, as they run a higher risk of confounding the data than probes that cross-hybridize to autosomal chromosomes.  

**snp_index** is a list of which probes do not contain SNPs. It is derived from fData(BTIC_GBM_Xeno)$n_SNPCpG, which is a matrix of how many SNPs are possible for that probe; if there's an NA that is good, because that means there's no SNP in that probe.  We then filter BTIC_GBM_Xeno, taking only the probes that were listed in **snp index**. We also remove probes which bind multiple locations in the genome as long as one of the locations is on the XY chromosome. The reason for this is as the X and Y chromosomes are not balanced amongst our samples (males vs females) we have to remove all probes which bind them so as to not skew the normalization. We do not remove multiple binders of probes which bind at various sites only in the autosomal chromosomal regions because they will most likely not skew our normalization and when we obtain our "hit list" we will verify them using pyrosequencing in order to determine if the specific site of interest is providing the signal we are seeing.

```{r echo=FALSE}
snp_index <- which(is.na(fData(BTIC_GBM_Xeno)$n_SNPCpG))
(n.polymorphic.probes<-(length(featureNames(BTIC_GBM_Xeno))-length(snp_index))) # 20,065
BTIC_GBM_Xeno <- BTIC_GBM_Xeno[snp_index, ] 
dim(BTIC_GBM_Xeno) # probes = 449,349  n = 12
xy_hit_index <- which(fData(BTIC_GBM_Xeno)$XY_Hits == "XY_NO")
(n.XYcrosshybrid.probes<-(length(featureNames(BTIC_GBM_Xeno))-length(xy_hit_index))) # 10597
BTIC_GBM_Xeno <- BTIC_GBM_Xeno[xy_hit_index, ] 
dim(BTIC_GBM_Xeno) # probes = 438,752, n = 12
# save(BTIC_GBM_Xeno, file = "BTIC_GBM_Xeno(filtered).RData")
```

We have removed 30,662 (20,065 polymorphic probes plus 10,597 XY cross-hybridizing probes) probes using the two filters described above. This leaves us with 438,752 probes for our analysis.

We will produce one more MDS plot and correlation heatmap using fully filtered (colour-corrected/background-subtracted) data.

```{r fig.width=8, fig.height=7, echo=FALSE}
tissueCol <- as.numeric(factor(BTIC_GBM_Xeno$Sample_Group))
tissueCol <- gsub("1", "red",  gsub("2", "forestgreen", gsub("3", "blue", tissueCol)))

par(cex = 0.7)
plotSampleRelation(BTIC_GBM_Xeno, method = "mds", col = tissueCol)
legend("top", legend = c("BTIC", "Tumour", "Xeno"),fill = c("red", "forestgreen", "blue"), horiz = T)
```



```{r fig.width=12, fig.height=12}
SexCol <- as.numeric(factor(BTIC_GBM_Xeno$Sex))
SexCol <-  gsub("1", "deeppink", gsub("2", "dodgerblue", SexCol))

BuPu <- colorRampPalette(brewer.pal(n = 9, "BuPu"))

cor <- cor(betas(BTIC_GBM_Xeno), use = "pairwise.complete.obs")

heatmap.2(cor,main = "Correlation of samples - 438,752 probes",
          trace = "none", col = BuPu(256), dendrogram = "both",
          RowSideColors = SexCol, cexRow = 0.8,
          ColSideColors = tissueCol, cexCol = 0.8,
          keysize = 1)
legend("top", legend = c("BTIC", "Tumour", "Xeno"),fill = c("red", "forestgreen", "blue"), horiz = T)
legend("top", legend = c("Female", "Male"),fill = c("deeppink","dodgerblue"), horiz = T)
```



### Summary table of probe/sample filtering processes:
Here is a table summarizing our probe filtering and pre-processing steps undertaken so far:

```{r echo=FALSE}
Filtered <- matrix(c("485,577", "12","Starting Probes", "485,512", "12","SNP Probes (65)", "473,864", "12","XY Probes (11,648)", "469,414", "12","Bad Probes (4450)", "449,349", "12","Polymorphic Probes (20,065)", "438,752", "12","XY Cross-hybridizing (10,597)", "438,752", "12", "Final Filtered Dataset"),ncol=3,byrow=TRUE)
colnames(Filtered) <- c("Probes","Samples","Filtering")
rownames(Filtered) <- c("I","II","III", "IV", "V", "VI", "VII")
Filtered <- as.table(Filtered)
print(Filtered)
```

```{r results='asis', echo=FALSE}
print(xtable(Filtered, align=c("llcr")), type="html")
```

We will continue with the normalization process using the objects BTIC_GBM_Xeno.RData and BTIC_GBM_Xeno.2.RData

### G. Normalization

Before we get into colour correction and normalization, we'll generate a density plot of the raw data from  with the probes filtered. We will accordingly compare these raw density plots to the normalized density plots at the end of the script.

```{r BTIC_GBM_Xeno_filteredDensity, echo=FALSE}
BatchCol <- rainbow(nlevels(factor(BTIC_GBM_Xeno$Batch)))[as.numeric(factor(BTIC_GBM_Xeno$Batch))]

density(BTIC_GBM_Xeno, col=BatchCol, lty = 1, legend=NULL, main = "BTIC_GBM_Xeno after probe filtering - M-values by Chip")
legend("topright",title="Batch", fill=rainbow(nlevels(factor(BTIC_GBM_Xeno$Batch))),legend=levels(factor(BTIC_GBM_Xeno$Batch)),cex=0.6)
```


```{r BTIC_GBM_Xeno.rev_tissueDensity, echo=FALSE}
tissueCol <- as.numeric(factor(BTIC_GBM_Xeno$Sample_Group))
tissueCol <- gsub("1", "red",  gsub("2", "forestgreen", gsub("3", "blue", tissueCol)))
density(BTIC_GBM_Xeno, col=tissueCol, lty = 1, legend=NULL, main = "BTIC_GBM_Xeno after probe filtering- M-values by tissue")
legend("topleft", legend = c("BTIC", "Tumour", "Xeno"),fill = c("red", "forestgreen", "blue"), horiz = T, cex=0.5)
```


```{r BTIC_GBM_Xeno.rev_SexDensity, echo=FALSE}
SexCol <- as.numeric(factor(BTIC_GBM_Xeno$Sex))
SexCol <- gsub("1", "deeppink", gsub("2", "dodgerblue", SexCol))
density(BTIC_GBM_Xeno, col=SexCol, lty = 1, legend=NULL, main = "BTIC_GBM_Xeno after probe filtering- M-values by Sex")
legend("topleft", legend = c("Female", "Male"),fill = c("deeppink","dodgerblue"), horiz = T, cex=0.7)
```
We observe limited variability in the methylation profiles that are driven by sex differences.

### H. Quantro

The quantro R-package can be used to test for global differences between groups of distributions which assess whether global normalization methods such as quantile normalization should be applied.

```{r}
# clone original filtered methylumi objects
all.dat<-BTIC_GBM_Xeno
all.dat.2<-BTIC_GBM_Xeno.2
library(quantro)
matboxplot(as.matrix(betas(all.dat)), groupFactor = all.dat$Sample_Group, xaxt = "n",main = "Distribution of Methylation Values in All.dat", ylab="Beta Value", xlab="Samples") 
```


```{r}
library(doParallel)
registerDoParallel(cores=4)

# all.dat
(qtestPerm.all <- quantro(as.matrix(betas(all.dat)), all.dat$Sample_Group, B = 1000)) # quantroStat = 3.233, quantroPvalPerm = 0.078
quantroPlot(qtestPerm.all)

#p<0.01 then there are global differences in distributions across groups
```

We can see that the quantro stat is not significant for the all.dat (permuted p-value = 0.078), so we can conclude that quantile normalization is appropriate for these datasets. 

### I. Quantile Normalization

```{r}
library(lumi)
all.dat.q <- lumiMethyN(all.dat, method = "quantile")
```


### J. Probe-type Normalization: BMIQ

The 450K Illumina Infinium Array has inherent variation associated with its methodologies which must be accounted for in our analyses. Much of this variation is attributed to the use of two types of probes used in the array, Type I and Type II.

Type I probes contain two bead types corresponding to an unmethylated (U) or methylated (M) status. Type I probes obtain methylation status as a result of fluoresence expressed after a single base pair extension occurs just after the target basepair, resulting in only one color channel being utilized (red). Type I probes also assume that any CpG sites underlying the probe are of the same status as the query site (methylated or unmethylated). The beta values for Type I probes are then determined by this formula b= M/(U + M). Importantly, Type I probes are enriched in regions of high CpG density (carry 3 or more CpG sites underlying the probe body), particularly those associated with promoters). Type II probes tend to occur in lower CpG density regions (carry 3 or less CpG sites underlying the probe body). Type II probes do not assume the methylation status of underlying CpG sites within the probe and so consist of a combination of degenerate probes with varying combinations of up to three underlying CpG sites. Type II probes also detect methylation status with a single base pair extension except that the site being extended is the CpG site of detection and so require two fluorescent colors green for methylated (M) and red for unmethylated (U) sites. Type II probe beta values are calculated using this formula b = Green (M)/(Red (U) + Green (M)). In terms of considerations for normalization, Type I probes have a much higher dynamic range than Type II probes, which may introduce an enrichment bias towards Type I probes when ranking probes in supervised analyses (Teschendorff et al 2013, Bioinformatics). 

Due to these inherent differences between Type I and Type II probes used in the Illumina Infinium 450K array several groups in the field have deveolped various normalization analyses to correct for the differences between these probes. We will be using an intra-sample normalization method that corrects for probe-type differences called BMIQ (Beta MIxture Quantile dilation) (Teschendorff et al 2013, Bioinformatics).

#### BMIQ (Beta Mixture Quantile dilation)

BMIQ is an intra-sample normalisation procedure, correcting the bias of type-2 probe values. BMIQ uses a 3-step procedure: (i) fitting of a 3-state beta mixture model, (ii) transformation of state-membership probabilities of type2 probes into quantiles of the type1 distribution, and (iii) a conformal transformation for the hemi-methylated probes. Exact details can be found in the reference (Teschendorff et al 2013, Bioinformatics).


```{r BTIC_GBM_Xeno.rev_BMIQ, echo=FALSE}
# define probe design vector
head(probe_design<-as.character(fData(all.dat.q)$INFINIUM_DESIGN_TYPE))
probe_design.v<- replace(probe_design, probe_design=="I", 1)
probe_design.cor<- replace(probe_design.v, probe_design.v=="II", 2)
probe_design.cor<-as.numeric(probe_design.cor)
identical(nrow(exprs(all.dat)), length(probe_design.cor))

# Run BMIQ across each dataset (this code takes a long time---best to run it overnight esp if you have a lot of samples)
all.dat_normalized<-apply(betas(all.dat.q), 2, function(x) BMIQ(x,probe_design.cor)) # all.dat


# extract normalized beta values and reshape
reshape.BMIQ.betas<-function(bmiq.output,methy.object){
  bb<-lapply(1:length(bmiq.output), function(x) bmiq.output[[x]]$nbeta)
  cc<-do.call(rbind, bb)
  dd<-t(cc)
  colnames(dd)<-colnames(exprs(methy.object))
  ee<-methy.object
  mvals<-Beta2M(dd)
  exprs(ee)<-mvals
  return(ee)
}

## reshape BMIQ output into new methylumi object for each dataset
BTIC_GBM_Xeno.BMIQ<-reshape.BMIQ.betas(all.dat_normalized,all.dat.q)

# save(BTIC_GBM_Xeno.BMIQ, file = "BTIC_GBM_Xeno.BMIQ.RData")
```

### K. Plot beta values before/after normalization

We will show the distribution of beta values for the entire dataset (all samples normalized together) in its raw form and final normalized form.

```{r BTIC_GBM_Xeno.rev_betaPlot, echo=FALSE} 

par(mfrow = c(1, 2))
density(BTIC_GBM_Xeno,legend=NULL, lty = 1, xlim = c(-6, 6), ylim=c(0,0.2), main = "Raw", col=tissueCol)
density(BTIC_GBM_Xeno.BMIQ,legend=NULL, lty = 1, xlim = c(-6, 6), ylim=c(0,0.2), main = "BMIQ", col=tissueCol)

type <- fData(BTIC_GBM_Xeno.BMIQ)$INFINIUM_DESIGN_TYPE

# for cancer.dat
par(mfrow = c(1, 1))
plot(c(0, 1) , c(0, 10), type = "n", xlab = "Beta values", ylab = "Density")
lines(density(na.omit(betas(BTIC_GBM_Xeno))[type=="I",]),col="blue",lty=1)
lines(density(na.omit(betas(BTIC_GBM_Xeno))[type=="II",]),col="blue",lty=2, wld=1.5)
lines(density(na.omit(betas(BTIC_GBM_Xeno.BMIQ))[type=="I",]),col="red",lty=1)
lines(density(na.omit(betas(BTIC_GBM_Xeno.BMIQ))[type=="II",]),col="red",lty=2, lwd=1.5)
legend("top", inset=.05, cex=0.6, c("Raw Type 1","Raw Type 2", "Normalized Type 1", "Normalized Type 2"), col=c("blue","blue","red", "red"), lty=c(1,2,1,2), horiz=FALSE)


```

We can see that there is improved peak-to-peak overlap of the Type 1 and Type 2 probes with the final normalized datasets as desired (ie Type 1 and 2 probes have more similar dynamic ranges for their beta values). 

We will continue on with our analysis using the BMIQ-normalized datasets.

### L. PCA to assess sources of variability in BTIC_GBM_Xeno data

We will perform PCA to assess sources of variability in the normalized BTIC_GBM_Xeno DNA methylation data 

## Heat scree plot Function
```{r}
### Function of association meta variable with PC (ANOVA)
heat_scree_plot<-function(Loadings, Importance, Num, Order){
  adjust<-1-Importance[1]
  pca_adjusted<-Importance[2:length(Importance)]/adjust
  pca_df<-data.frame(adjusted_variance=pca_adjusted, PC=seq(1:length(pca_adjusted)))
  
  scree<-ggplot(pca_df[which(pca_df$PC<Num),],aes(PC,adjusted_variance))+geom_bar(stat = "identity",color="black",fill="grey")+theme_bw()+
        theme(axis.text = element_text(size =12),
              axis.title = element_text(size =15),
              plot.margin=unit(c(1,1.5,0.2,2.25),"cm"))+ylab("Variance")+
    scale_x_continuous(breaks = seq(1,Num,1))
  
  #### Heat
  ## correlate meta with PCS
  ## Run anova of each PC on each meta data variable

  aov_PC_meta<-lapply(1:ncol(meta_categorical), function(covar) sapply(1:ncol(Loadings), function(PC) summary(aov(Loadings[,PC]~meta_categorical[,covar]))[[1]]$"Pr(>F)"[1]))
  cor_PC_meta<-lapply(1:ncol(meta_continuous), function(covar) sapply(1:ncol(Loadings), function(PC) (cor.test(Loadings[,PC],as.numeric(meta_continuous[,covar]),alternative = "two.sided", method="spearman", na.action=na.omit)$p.value)))
  names(aov_PC_meta)<-colnames(meta_categorical)
  names(cor_PC_meta)<-colnames(meta_continuous)
  aov_PC_meta<-do.call(rbind, aov_PC_meta)
 cor_PC_meta<-do.call(rbind, cor_PC_meta)
 aov_PC_meta<-rbind(aov_PC_meta, cor_PC_meta)
  aov_PC_meta<-as.data.frame(aov_PC_meta)
  #adjust
  aov_PC_meta_adjust<-aov_PC_meta[,2:ncol(aov_PC_meta)]
  
    
  #reshape
  avo<-aov_PC_meta_adjust[,1:(Num-1)]
  avo_heat_num<-apply(avo,2, as.numeric)
  avo_heat<-as.data.frame(avo_heat_num)
  colnames(avo_heat)<-sapply(1:(Num-1), function(x) paste("PC",x, sep=""))
  avo_heat$meta<-rownames(avo)
  avo_heat_melt<-melt(avo_heat, id=c("meta"))
  
  # cluster meta data
  ord <- Order
  meta_var_order<-unique(avo_heat_melt$meta)[rev(ord)]
  avo_heat_melt$meta <- factor(avo_heat_melt$meta, levels = meta_var_order)
  
  # color if sig
  avo_heat_melt$Pvalue<-sapply(1:nrow(avo_heat_melt), function(x) if(avo_heat_melt$value[x]>=0.9){">=0.9"}else{
  if(avo_heat_melt$value[x]>=0.5){">=0.5"}else{
  if(avo_heat_melt$value[x]>=0.1){">=0.1"}else{"<0.1"}}})
  avo_heat_melt$Pvalue<-sapply(1:nrow(avo_heat_melt), function(x) if(avo_heat_melt$value[x]<=0.001){"<=0.001"}else{
     if(avo_heat_melt$value[x]<=0.01){"<=0.01"}else{
       if(avo_heat_melt$value[x]<=0.05){"<=0.05"}else{">0.05"}}})
  
  heat<-ggplot(avo_heat_melt, aes(variable,meta, fill = Pvalue)) +
  geom_tile(color = "black",size=0.5) +
  theme_gray(8)+scale_fill_manual(values=c("#084594","#4292c6","#9ecae1","#deebf7"))+
      theme(axis.text = element_text(size =10, color="black"),
            axis.text.x = element_text(),
          axis.title = element_text(size =15),
          legend.text = element_text(size =14),
          legend.title = element_text(size =12),
          legend.position = c(1, 0), legend.justification = c(1,0),
          plot.margin=unit(c(0,2.25,1,1),"cm"))+
    xlab("Principal Component")+ylab(NULL)
  
  grid.arrange(scree, heat, ncol=1)
}
```


```{r}

dim(BTIC_GBM_Xeno.BMIQ) # probes = 438,752, samples = 12
ind<-is.row.na(betas(BTIC_GBM_Xeno.BMIQ)) # The function returns a vector of logical variables, one for each row of the matrix. The variable is TRUE if the row does not contain any missing values and FAlSE otherwise.
length(na.count<-which(ind=="FALSE")) # 0 (there are no rows that contain a NA in the raw data)
```


## PCA Scree Heatmap for BTIC_GBM_Xeno.BMIQ

```{r warning=FALSE, fig.height=9, fig.width=11}

## re-structure meta data: change categorical variables to factors for ANOVA and continuous variables to numeric for Spearman's correlation
meta<-pData(BTIC_GBM_Xeno.BMIQ)
str(meta)
meta$Tissue<-as.factor(meta$Tissue)
meta$Sentrix_ID<-as.factor(meta$Sentrix_ID)
meta$Diagnosis<-as.factor(meta$Diagnosis)
meta$Batch<-as.factor(meta$Batch)
meta$Row<-as.factor(meta$Row)
meta$Sex<-as.factor(meta$Sex)

## PCA
betas.BTX<-betas(BTIC_GBM_Xeno.BMIQ)
PCA_full<-princomp(betas.BTX[complete.cases(betas.BTX),])
Loadings<-as.data.frame(unclass(PCA_full$loadings))
vars <- PCA_full$sdev^2
Importance<-vars/sum(vars)
adjust<-1-Importance[1]
pca_adjusted<-Importance[2:length(Importance)]/adjust
(pca_df<-data.frame(adjusted_variance=pca_adjusted, PC=seq(1:length(pca_adjusted))))
sum(pca_df$adjusted_variance[1:7]) # top 7 PCs account for ~91% of the DNA methylation variance

#Specify which covariates are categorical and/or categorical
colnames(meta)

meta_categorical<-meta[,c("Tissue","Diagnosis", "Batch", "Sentrix_ID", "Row", "Sex")]  # input column numbers in meta that contain categorical variables; should all be structured as factor
meta_continuous<-meta[,c("Age")]  # input column numbers in meta that contain categorical variables; should all be structured as factor
meta_continuous<-data.frame(meta_continuous)
colnames(meta_categorical)<-c("Source","Diagnosis","Batch", "Chip", "Row", "Sex")
colnames(meta_continuous)<-c("Age")
# Specify the number of PCs you want shown (usually # of samples in the dataset)
Num<-11
# Designate what order you want the variables to appear (continuous variables rbinded to categorical variables in function)
Order<-1:7

#Apply function on PCA results, pulls in the meta data and beta values from above
heat_scree_plot(Loadings, Importance, Num, Order)  
```

We can see that various tested variables appear to contribute to the DNA methylation variability in the normalized BTIC_GBM_Xeno object. Specifically, Diagnosis correlates with PC4, Batch correlates with PC5, Chip associates with PC9, Row associates with PC1, Sex associates with PC2 and Age associates with PC3. Oddly, Source (ie BTIC, Tumour or Xeno) does not associate with any of the PCs. However, given that PCA or other forms of factor analysis become unstable with low sample sizes (Mundfrom, D.J., Shaw, D.G., & Ke, T.L. (2005). Minimum sample size recommendations for conducting factor analyses. International Journal of Testing, 5 (2), 159-168.), we will have to be somewhat weary of these results and visually assess for each variable/PC combo individually (se we do below).

```{r}
# ggplot of PC1 vs PC2 for row
### in ggplot need to adjust for probe offset PC so plot Comp.2 vs Comp.3 to represent PC1 vs PC2
Row <-factor(BTIC_GBM_Xeno.BMIQ$Subtype)
ggplot(Loadings,aes(Comp.2,Comp.3))+
  geom_point(aes(fill=Row),shape=21,size=5)+
  theme_bw() +
  xlab("PC1") + 
  ylab("PC2")
## Do not observe a distinct effect of row on PC1

# ggplot of PC1 vs PC2 for sex
### in ggplot need to adjust for probe offset PC so plot Comp.2 vs Comp.3 to represent PC1 vs PC2
Sex <-factor(BTIC_GBM_Xeno.BMIQ$Sex)
ggplot(Loadings,aes(Comp.2,Comp.3))+
  geom_point(aes(fill=Sex),shape=21,size=5)+
  theme_bw() +
  xlab("PC1") + 
  ylab("PC2")
## Observe a distinct effect of sex on PC2 (however, with only 1 female represented in 3 matched samples, this result is likely not very robust). 

# ggplot of PC2 vs PC3 for age
### in ggplot need to adjust for probe offset PC so plot Comp.2 vs Comp.3 to represent PC1 vs PC2
Age <-as.numeric(BTIC_GBM_Xeno.BMIQ$Age)
ggplot(Loadings,aes(Comp.4,Comp.5))+
  geom_point(aes(fill=Age),shape=21,size=5)+
  theme_bw() +
  xlab("PC3") + 
  ylab("PC4")
## Do not observe a distinct effect of age on PC3

# ggplot of PC3 vs PC4 for diagnosis
### in ggplot need to adjust for probe offset PC so plot Comp.2 vs Comp.3 to represent PC1 vs PC2
Diagnosis <-factor(BTIC_GBM_Xeno.BMIQ$Diagnosis)
ggplot(Loadings,aes(Comp.4,Comp.5))+
  geom_point(aes(fill=Diagnosis),shape=21,size=5)+
  theme_bw() +
  xlab("PC3") + 
  ylab("PC4")
## Do observe a reasonable association between PC4 and Diagnosis

# ggplot of PC4 vs PC5 for batch
### in ggplot need to adjust for probe offset PC so plot Comp.2 vs Comp.3 to represent PC1 vs PC2
Batch <-factor(BTIC_GBM_Xeno.BMIQ$Batch)
ggplot(Loadings,aes(Comp.5,Comp.6))+
  geom_point(aes(fill=Batch),shape=21,size=5)+
  theme_bw() +
  xlab("PC4") + 
  ylab("PC5")
## Do observe a somewhat association between PC5 and Batch. However Batch is confounded with Tissue source as all Xenografts were ran on the second batch. 

# ggplot of PC8 vs PC9 for chip
### in ggplot need to adjust for probe offset PC so plot Comp.2 vs Comp.3 to represent PC1 vs PC2
Chip <-factor(BTIC_GBM_Xeno.BMIQ$Sentrix_ID)
ggplot(Loadings,aes(Comp.9,Comp.10))+
  geom_point(aes(fill=Chip),shape=21,size=5)+
  theme_bw() +
  xlab("PC8") + 
  ylab("PC9")
## Do not observe a distinct effect of chip on PC9

# ggplot of PC1 vs PC2 for Tissue source
### in ggplot need to adjust for probe offset PC so plot Comp.2 vs Comp.3 to represent PC1 vs PC2
Tissue <-factor(BTIC_GBM_Xeno.BMIQ$Sample_Group)
ggplot(Loadings,aes(Comp.2,Comp.3))+
  geom_point(aes(fill=Tissue),shape=21,size=5)+
  theme_bw() +
  xlab("PC1") + 
  ylab("PC2")
## Observe a subtle effect of Tissue Source on PC1; will statistically test this potential association

source.PC1.df<-data.frame(Sample = rownames(meta), Tissue = as.factor(meta$Tissue), PC1_Score = Loadings$Comp.2)

kruskal.test(PC1_Score ~ Tissue, data = source.PC1.df) # p = 0.031 (significant association between PC1 and Tissue source)


diagnosis.PC4.df<-data.frame(Sample = rownames(meta), Diagnosis = as.factor(meta$Diagnosis), PC4_Score = Loadings$Comp.5)

kruskal.test(PC4_Score ~ Diagnosis, data = diagnosis.PC4.df) # p = 0.025 (significant association between PC4 and Diagnosis)
```

We observe significant association of PC1 and tissue source (p = 0.031, Kruskal Wallis test) and PC4 and Diagnosis (p = 0.025, Kruskal Wallis test)

### M. Hierarchical cluster analysis

We will perform unsupervised hierarchical cluster analysis to analyze relationships between global DNA methylation profiles between the samples in BTIC_GBM_Xeno

```{r}

load("BTIC_GBM_Xeno.BMIQ.RData")
## Heatmap
# define colour palette
BuPu <- colorRampPalette(brewer.pal(10,"RdYlBu")[1:10])(30)
cor <- cor(betas(BTIC_GBM_Xeno.BMIQ), use = "pairwise.complete.obs")
# define coloured variables
str(pData(BTIC_GBM_Xeno.BMIQ))
colnames(pData(BTIC_GBM_Xeno.BMIQ))
tissueCol <- as.numeric(factor(BTIC_GBM_Xeno.BMIQ$Tissue))
tissueCol <- gsub("1", "#fc8d59",  gsub("2", "#99d594", gsub("3", "#bcbddc", tissueCol)))
diagnosisCol <- as.numeric(factor(BTIC_GBM_Xeno.BMIQ$Diagnosis))
diagnosisCol <- gsub("1", "aquamarine4",  gsub("2", "hotpink2", diagnosisCol))
heatmap.2(cor, key = T,
          trace = "none", col = BuPu, dendrogram = "both",
           cexRow = 1,
          ColSideColors = tissueCol,
          RowSideColors = diagnosisCol,
          cexCol = 1,
          keysize = 0.75, key.par = list(cex = 0.5))
legend("topright",legend = c("cell", "tumour", "xeno"), 
    col = c("#99d594", "#fc8d59", "#bcbddc"), 
    lty= 1,             
    lwd = 8            
)

heatmap.2(cor, key = T,
          trace = "none", col = BuPu, dendrogram = "both",
           cexRow = 1,
          ColSideColors = tissueCol,
          cexCol = 1,
          keysize = 0.75, key.par = list(cex = 0.5))
legend("topright",legend = c("cell", "tumour", "xeno"), 
    col = c("#99d594", "#fc8d59", "#bcbddc"), 
    lty= 1,             
    lwd = 8            
)
        
```

Hierachical cluster analysis shows that xenograft and BTICs largely cluster together, apart from the tumour. However in the case of 147_xeno, it clusters closer to 147_tumour over 147_cell. This may in part be explained by the 